src/Confluent.Kafka/Diagnostics.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright 2016-2018 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Confluent.Kafka
+{
+    internal static class Diagnostics
+    {
+        public const string DiagnosticListenerName = "Confluent.Kafka.Listener";
+        public const string TraceParentHeaderName = "traceparent";
+        public const string TraceStateHeaderName = "tracestate";
+
+        public static DiagnosticListener DiagnosticSource { get; } = new DiagnosticListener(DiagnosticListenerName);
+
+        public static class Producer
+        {
+            public const string ActivityName = DiagnosticListenerName + ".MessageProduced";
+            public const string StartName = ActivityName + ".Start";
+            public const string StopName = ActivityName + ".Stop";
+            public const string ExceptionName = ActivityName + ".Exception";
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Activity Start<TKey, TValue>(string topic, Message<TKey, TValue> message)
+            {
+                if (DiagnosticSource.IsEnabled(ActivityName, topic))
```

Comment:
The new recommendation from .NET would be to use `ActivitySource` and `StartActivity`, from `System.Diagnostics.DiagnosticSource` package 5.0.0.
https://github.com/dotnet/designs/pull/98/files#diff-25bba2a21bc64ddeccd374c1c28a0fefR24

DiagnosticSource package supporting the new pattern is releasing stable in Nov 2020, but previews will be available this week onwards.

Should we wait for new diagnosticsource to be available, and then instrument using the new approach? 
If instrumented with new approach, then there wouldn't be any need of adapters in OpenTelemetry. 
@reyang also to share thoughts on this.